### PR TITLE
Fact sheet refactor

### LINKED ISSUE #538: Name of submissions: missing?
We need to be able to add a meaningful name to submissions (in addition to the automatic labels 'File Name' and 'Username'). This name should appear on the leaderboard (in addition to or instead of Username)

### LINKED ISSUE #550: Refactor Fact_Sheet Question/Answer system

### PR DESCRIPTION
**Pull request by Logan-Ruf**
# @ mention of reviewers
@ckcollab 
@gibsonbailey 

# Screenshots
Edit Form
image
Submission Upload
image
Submission Details
image
Leaderboard
image

# A brief description of the purpose of the changes contained in this PR.
Update to the Fact_Sheet system. Adds metadata for `is_required`, `is_on_leaderboard`, and `title`
`is_required` for selection questions will create a blank default entry
`is_required` for text entry does not validate empty strings
`is_on_leaderboard` will show the answer to a question on the leaderboard
`title` will be the prompt the user sees on the fact_sheet and on the leaderboard

# Issues this PR resolves
Resolves #550
Resolves #538

# A checklist for hand testing
- [x] Uploading a submission and adding to a leaderboard without a fact_sheet should work as intended
- [x] Uploading a submission with a fact_sheet question marked as `is_on_leaderboard` should show answer on the leaderboard
- [x] A selection question not marked  `is_required` should have a blank option
- [x] A text question marked as `is_required` should not validate when submitting whitespace or nothing

# Checklist
- [x] Code review by me 
- [x] Hand tested by me 
- [x] I'm proud of my work
- [ ] Code review by reviewer
- [ ] Hand tested by reviewer
- [ ] Ready to merge

### CODE REVIEW

File: src/apps/api/serializers/submissions.py [hunk 1]
```diff
@@ -142,11 +143,13 @@ def validate(self, attrs):
             fact_sheet = data['phase'].competition.fact_sheet
             if set(fact_sheet_answers.keys()) != set(fact_sheet.keys()):
                 raise ValidationError("Fact Sheet keys do not match Answer keys")
-            for key in fact_sheet_answers.keys():
+            for key in fact_sheet_answers:
```
Comment (ckcollab) [inline src/apps/api/serializers/submissions.py:146]: you could do..
```python
for key, value in fact_sheet_answers.items():
```

then you don't have to dereference `fact_sheet_answers[key]` over and over, can use `value`

File: src/apps/competitions/tests/test_submissions.py [hunk 2]
```diff
@@ -246,27 +246,61 @@ class FactSheetTests(SubmissionTestCase):
...
+                "selection": ["", "v1", "v2", "v3"],
+                "is_required": "false",
+                "is_on_leaderboard": "true"
+            }
         }
         self.competition.save()
 
     def test_fact_sheet_valid(self):
         submission = SubmissionCreationSerializer(super().make_submission()).data
+        from pprint import pprint
```
Comment (ckcollab) [inline src/apps/competitions/tests/test_submissions.py:286]: pull prints?

File: src/apps/competitions/tests/test_submissions.py [hunk 3]
```diff
@@ -246,27 +246,61 @@ class FactSheetTests(SubmissionTestCase):
...
         submission['fact_sheet_answers'] = {
             "boolean": True,
-            "selection": "value3",
-            "text": "accept any",
+            "selection": "v3",
+            "text_required": "accept_text",
+            "text": "",
         }
-        serializer = SubmissionCreationSerializer(data=submission, instance="PATCH")
+        serializer = SubmissionCreationSerializer(instance='PATCH', data=submission)
```
Comment (ckcollab) [inline src/apps/competitions/tests/test_submissions.py:294]: what is this attempting to do? instance should be an instance of a model?

File: src/static/riot/competitions/detail/leaderboards.tag [hunk 4]
```diff
@@ -101,6 +108,21 @@
                 .done(responseData => {
                     self.selected_leaderboard = responseData
                     self.columns = []
+                    if(self.selected_leaderboard.fact_sheet_keys){
+                        let fake_metadata_task = {
```
Comment (ckcollab) [inline src/static/riot/competitions/detail/leaderboards.tag:112]: maybe add a comment about this section's purpose?

File: src/apps/competitions/tests/test_submissions.py (hunk 3 above)
Comment (Logan-Ruf) [reply src/apps/competitions/tests/test_submissions.py:294]: I'm not a 100%. I got it from a stack overflow somewhere. If instance='PATCH' or Submission the tests work as expected. If instance isn't defined, then the tests fail.

Comment (ckcollab) [reply src/apps/competitions/tests/test_submissions.py]: that's... weird, what happens if you pass `instance='ABC'`?

Comment (Logan-Ruf) [reply src/apps/competitions/tests/test_submissions.py]: That works as well. If I don't have anything I get this error:
image

Comment (ckcollab) [reply src/apps/competitions/tests/test_submissions.py]: I think instance should be a model instance, passing a string here definitely seems wrong..

File: src/static/riot/competitions/editor/_competition_details.tag [hunk 5]
```diff
@@ -225,13 +241,32 @@
...
+                }
+                for(entry of q_serialized){
+                    if(entry.name.split('-')[0] === 'selection') {
+                        let selection = entry.value.split(',')
+                        selection = selection.map(s => s.trim()).filter(s => s !== '')
+                        form_json[question_key][entry.name.split('-')[0]] = selection
+                    } else if (entry.name.split('-')[0] === 'key'){
+                        // Check to make sure key isn't empty
+                        if(!entry.value){
+                            console.log("returning false")
```
Comment (Logan-Ruf) [inline src/static/riot/competitions/editor/_competition_details.tag:260]:
```suggestion
```

File: src/static/riot/competitions/detail/leaderboards.tag [hunk 6]
```diff
@@ -111,7 +115,24 @@
             CODALAB.api.get_leaderboard_for_render(self.phase_id)
                 .done(responseData => {
                     self.selected_leaderboard = responseData
+                    console.log(responseData)
```
Comment (Logan-Ruf) [inline src/static/riot/competitions/detail/leaderboards.tag:118]:
```suggestion
```

File: src/apps/api/serializers/competitions.py [hunk 7]
```diff
@@ -135,6 +135,24 @@ def validate_phases(self, phases):
...
+        expected_keys = {"key", "type", "title", "selection", "is_required", "is_on_leaderboard"}
+        valid_question_types = {"checkbox", "text", "select"}
+        for key, value in fact_sheet.items():
+            missing_keys = expected_keys.symmetric_difference(set(value.keys()))
+            if missing_keys:
+                raise ValidationError(f'Missing {missing_keys} values for {key}')
+            if key != value['key']:
+                raise ValidationError(f"key:{value['key']}  does not match JSON key:{key}")
+            if value['type'] not in valid_question_types:
+                raise ValidationError(f"{value['type']} is not a valid question type")
```
Comment (ckcollab) [inline src/apps/api/serializers/competitions.py:153]: awesome validation

File: src/apps/api/tests/test_competitions.py [hunk 8]
```diff
@@ -240,3 +241,100 @@ def test_get_competition_leaderboard_as_zip(self):
...
+                "is_required": "false",
+                "is_on_leaderboard": "true"
+            }
+        }
+
+    def test_competition_working_fact_sheet(self):
+        new_comp_data = self.competition_data
+        new_comp_data['fact_sheet'] = self.competition_fact_sheet
+        competition_serializer = CompetitionSerializer(instance=self.competition, data=new_comp_data)
+        assert competition_serializer.is_valid()
```
Comment (ckcollab) [inline src/apps/api/tests/test_competitions.py:296]: can you assert for integrity here too, something like..

```python
competition_serializer.is_valid(raise_exceptions=True)  # should add raise_exceptions here!
comp = competition_serializer.save()
assert comp.fact_sheet == new_comp_data['fact_sheet'] 
```

**Comment by ckcollab** [comment]
image
may be nice to have some kind of tool tip or something explaining why we need a key/such?

**Comment by ckcollab** [comment]
ultra nitpick:
image

Meta-data -> Fact sheet answers?

**Pull request merged**